velox/functions/prestosql/window/Rank.cpp

```diff
@@ -41,25 +50,68 @@ class RankFunction : public exec::WindowFunction {
       vector_size_t resultOffset,
       const VectorPtr& result) override {
     int numRows = peerGroupStarts->size() / sizeof(vector_size_t);
-    auto* rawValues = result->asFlatVector<int64_t>()->mutableRawValues();
     auto* rawPeerStarts = peerGroupStarts->as<vector_size_t>();
+
+    void* rawValues = NULL;
+    switch (R) {
+      case RankType::kRank:
+      case RankType::kDenseRank: {
+        auto* resultRawValues =
+            result->asFlatVector<int64_t>()->mutableRawValues();
+        rawValues = resultRawValues;
+        break;
+      }
+      case RankType::kPercentRank: {
+        auto* resultRawValues =
+            result->asFlatVector<double>()->mutableRawValues();
+        rawValues = resultRawValues;
+        break;
+      }
+    }
+
     for (int i = 0; i < numRows; i++) {
       auto start = rawPeerStarts[i];
-      if (start != currentPeerGroupStart_) {
-        currentPeerGroupStart_ = start;
-        rank_ += previousPeerCount_;
-        previousPeerCount_ = 0;
+      switch (R) {
+        case RankType::kRank: {
+          if (start != currentPeerGroupStart_) {
+            currentPeerGroupStart_ = start;
```

Comment:
These 3 lines are common across all the cases. You can move those lines before the switch statement, so that they appear only once.

Reply:
Thanks for the suggestion, the denseRank case was different earlier but have clubbed it now.